# ðŸ§« scanpy_tutorial
Identifying and visualizing marker genes with Scanpy

RNA-seq is a technique used to analyze the expression of genes in a biological sample that works by converting RNA molecules into complementary DNA (cDNA), sequencing this cDNA, and mapping the sequence data to a reference genome transcriptome to quantify gene expression levels. 

Whereas traditional bulk RNA sequencing provides an average gene expression profile for a population of cells, single-cell RNA-seq (scRNA-seq) allows for the study of gene expression at a single-cell level, providing insights into cellular heterogeneity and gene expression patterns within individual cells. As a result, scRNA-seq has diverse applications in studying disease mechanisms in specific cell types, identifying new drug targets, and discovering new biomarkers. 


# ðŸ§« Analyzing scRNA-Seq Data With ScanPy

Scanpy is a Python package that analyzes single-cell RNA sequencing (scRNA-seq) data. It provides a wide range of functionalities for preprocessing, analyzing, and visualizing scRNA-seq data, making it a powerful tool for researchers studying the complex biology of individual cells. While it is possible to perform scRNA-seq analysis without Scanpy, using the package can streamline the process and provide access to a range of advanced analysis tools and techniques, which is especially helpful for researchers new to single-cell genomics.

In this tutorial, I will demonstrate some of ScanPy's core functionalities using a reduced version of the 68k PBMC (peripheral blood mononuclear cell) dataset generated by the 10x Genomics platform. The reduced version of this dataset consists of 700 cells and 765 highly variable genes, making it easier for beginners like myself to analyze.

Importantly, this reduced dataset is preprocessed already, so we don't need to worry about quality control, filtering out low-quality cells and genes, normalizing counts, or batch correction. Additionally, UMAP dimensionality reduction has already been used on our dataset, allowing us to visualize the data in a lower-dimensional space while preserving its underlying structure, allowing us to explore the dataset right away.

Now, we can use the code below to setup our ScanPy work environment and load the PBMC dataset:
```
import scanpy as sc
sc.set_figure_params(dpi=100, color_map="viridis_r")
sc.settings.verbosity = 0
sc.logging.print_header()
pbmc = sc.datasets.pbmc68k_reduced()
pbmc
```
Below, youâ€™ll find a step-by-step breakdown of the code block above:
- **import scanpy** as sc imports the ScanPy package and allows you to access its functions and classes using the sc alias. 
- **sc.set_figure_params(dpi=100, color_map=â€™viridis_râ€™)** sets the parameters for the figures generated by ScanPy. For example, dpi=100 sets the resolution of figures to 100 dots per inch, and color_map=â€™viridis_râ€™ sets the default color map. 
- **sc.settings.verbosity=0** sets the verbosity level of ScanPy. By setting it to 0, only critical messages are displayed, which reduces the amount of output displayed during analysis.
- **sc.logging.print_header()** prints a header to the console, indicating that the ScanPy system is active. This header typically includes information about the version of Scanpy being used, as demonstrated in the image below:
- **pbmc = sc.datasets.pbmc68k_reduced()** loads a pre-processed single-cell RNA sequencing datasets called pbmc68k_reduced, which is a subset of the Peripheral Blood Mononuclear cells (PBMCs) dataset, which is commonly used for benchmarking scRNA-seq analysis methods. 
- **pbmc** then provides a comprehensive overview of the scRNA-seq dataset, including information about individual cells, genes, clustering results, and dimensionality reduction.

# ðŸ§« Identifying Clusters By Marker Genes With UMAP Plots
After loading our PBMC dataset, we can use ScanPyâ€™s built in functionality to visualize the expression of select marker genes that are highly expressed in specific cell types, making them useful for distinguishing between cell types based on their gene expression profiles. Below youâ€™ll find a list of the various marker genes weâ€™ll visualize in this section, followed by the cell types in the immune system that they are highly expressed in:
- **CD79A** and **MS4A1** are commonly used B-cell marker genes.
- **IGJ (JCHAIN)** is a marked gene that is highly expressed in plasma cells, which are a type of B-cell that produces antibodies.
- **CD3D** is a marker gene that is used to identify T-cells, which is a type of immune cell involved in cell-mediated immunity.
- **GNLY** and **NKG7** are marker genes that are often used to identify natural killer (NK) cells, which are a type of cytotoxic lymphocyte that plays a row in the immune response.
- **CST3** and **LYZ** are marker genes that are commonly used to identify myeloid cells, which are a group of immune cells that include monocytes, macrophages, neutrophils, and dendritic cells.
- **FCGR3A** is a marker gene that is often used to identify monocytes, which are a type of white blood cell that can differentiate into macrophages or dendritic cells.
- **FCER1A** is a marker gene that is commonly used to identify dendritic cells, which are antigen-presenting cells that play a key role in the immune response.

Now, to visualize the marker genes above, we will use a UMAP plot, which is a type of scatter plot that visualizes the high-dimensional gene expression data of individual cells in a lower-dimensional space while preserving the local and global structure of the data. In a UMAP plot, each point represents a single cell, and the position of the point in the plot is determined by the reduced dimensions calculated by the UMAP algorithm. Cells that are close together in the UMAP plot are more similar in their gene expression profiles, while far apart cells are more dissimilar. UMAP plots are commonly used in scRNA-seq analysis to visualize cell clusters, identify cell types, and explore the heterogeneity of cell populations. In the code below, Iâ€™ll demonstrate how to create a UMAP plot:

```
color_vars = ['CD79A','MS4A1','IGJ','CD3D','FCER1A','FCGR3A','n_counts', 'bulk_labels']
sc.pl.umap(pbmc, color=color_vars, s=50, ncols=3, vmax="p99")
```
The code block above uses ScanPy to create a UMAP plot with multiple color-coded annotations. Below, Iâ€™ll break the code down step by step:

- **color_vars = ['CD79A','MS4A1','IGJ','CD3D','FCER1A','FCGR3A','n_counts', 'bulk_labels']** is a list containing the names of our genes and other variables (i.e., n_counts and bulk_labels) that will be used to color the points on the UMAP plot. 
- **sc.pl.umap(pbmc, color=color_vars, s=50, ncols=3, vmax="p99")** creates a UMAP plot using ScanPyâ€™s umap() function. Within the umap() function, we have multiple arguments, including pbmc, which is our dataset containing ssRNA-seq data, color=color_vars, which specifies the variable to use for coloring the UMAP plot, as well as additional parameters that specify the size of the plot, the number of columns, and the maximum color scale value (ensuring outliers do not overly influence the color scale).
  
The resultant plot can be seen below:
![umap plot](scanpy_tutorial/images/umap1.png)
